### DISCUSSION TITLE
[mini.icons] Adding icons like ansible to filetype

### DISCUSSION THREAD

**Discussion by mstinsky** [nvim-mini/mini.nvim#1017] [Q&A] [answered]
Greetings,

I just got mini.icons through the last update of lazyvim and would like to add icons to specific filetypes.
As a first test I wanted to add the ansible icons to filetype `yaml.ansible`.

What I did was adding the following to the lazyvim configuration:
```
return {
  {
    "echasnovski/mini.icons",
    opts = {
      filetype= {
        ['yaml.ansible'] = { glyph = '󱂚', hl = 'MiniIconsGrey' },
      },
    },
  },
}
```
Icon would be nf-md-ansible.

But for some reason that does not work for me. Doing the same for filetype `yaml` instead of `yaml.ansible` works, but then all my yaml files have the ansible icon which is not what I want to achieve.
It seems to me that for some reason mini.icons only sees the `yaml` filetype and not my `yaml.ansible`.

To detect the ansible filetype I am setting those through vim.add.filetype in a custom config in lazyvim. Which is correctly identified by the ansiblels.

Does anyone has a hint for me what I might be doing wrong?
Thanks in advance.

**Comment by echasnovski** [comment]
Hi!

First, just to make sure, is "yaml.ansible" is indeed a filetype? What it means is that when you open a target file, it has `'filetype'` options equal to `'yaml.ansimble'`. Does it say so after you execute `:=vim.bo.filetype`?

Second (if the first one is true), where do you expect it to be shown? If in the file explorer, then chances are that it uses "file" category. In this case, it has [certain order of preference](https://github.com/echasnovski/mini.nvim/blob/d187614ecee84ef8d7456339ed7ce1160d054310/doc/mini-icons.txt#L292-L296) and if the file name ends with plain "yaml" or "yml" extension, then it stops on the second step as those are manually [tracked extensions](https://github.com/echasnovski/mini.nvim/blob/d187614ecee84ef8d7456339ed7ce1160d054310/lua/mini/icons.lua#L658-L659) (you can see those with `:=MiniIcons.list('extension')`).

The reason it was added to manually tracked list is because it is fairly common (like, for example, around 3% of all files in [this benchmark project](https://github.com/Yggdroot/a_large_project)) and always relying on `vim.filetype.match()` might get slow-ish.

**Reply by mstinsky** [reply]
`set ft?` shows `filetype=yaml.ansible` the ansiblels is also active for my files so the filetype should be correct.

lazyvim is using neo-tree.nvim and thats where I am trying to get the correct icon. So I think your assumption could be correct as the file extension for ansible files are just `yml/yaml`. 

Is there any way that what I am trying to do might work?

**Reply by echasnovski** [reply]
I see.

> Is there any way that what I am trying to do might work?

Right now - only customizing "file" with explicit file names, but I imagine this won't work here.

Let me think about this, as similar situation has already popped up several times.

**Reply by mstinsky** [reply]
Sadly the default is that ansible files have the file extension `yml`.
I would say every existing role or collection has it like this, therefore changing the file extension is not feasible.

**Comment by echasnovski** [comment]
So, I've removed 'yml' and 'yaml' from the list of built-in (manually tracked) extensions, but upon investigation I don't think this will help your use case. Based on [this filetype detection](https://github.com/mfussenegger/nvim-ansible/blob/main/ftdetect/ansible.lua) for 'yaml.ansible' it requires checking full path, which is currently not supported: only basename is taken into account for "file" and "directory" category.

The reason for this is to make better use of caching: with full paths cache will not be reused for files with same basenames (which I think is a more common occurrence).

I'll make some profiling (both memory and speed) for when full path is used and get back to you. Hope to do this in the near future.

**Reply by mstinsky** [reply]
First of all thank you very much @echasnovski for taking your time and looking into this and trying to support my use case, very much appreciated!

I am not using the mentioned code from mfussenegger/nvim-ansible but one of the implementations in my config works very much the same with doing a regex over the file path. My config has a second mechanism where I do a regex over the first couple of lines of the buffer and doing a regex over the content to search for `regex = vim.regex "hosts:\\|tasks:"` and if found sets the `yaml.ansible` filetype.

Sadly as I have very little programming know-how and less so in lua I cant really help in figuring out a solution to this problem.

**Reply by echasnovski** [reply]
> Sadly as I have very little programming know-how and less so in lua I cant really help in figuring out a solution to this problem.

No worries whatsoever :) I am fairly certain that the issue here is that `vim.filetype.match()` is used only with basename and not full path (needed for proper detection of 'yaml.ansible' filetype). It works in quick local test. Now I need to figure out a reasonably good way for making it work with as less side effects as possible. If no such way is found, then this use case won't be supported, unfortunately. I'll do my best :)

**Comment by echasnovski** [comment, marked as answer]
@mstinsky, thanks to a series of refactors and optimizations (see [this comment](https://github.com/echasnovski/mini.nvim/issues/1007#issuecomment-2252493138) if you are curious which ones), this is now possible on latest `main` with a few necessary setup tweaks. Here is a setup that makes it work:

```lua
local ext_skip = { yml = true }
require('mini.icons').setup({
  use_file_extension = function(ext, _) return not ext_skip[ext:lower()] end,
})
```

This setup basically tells 'mini.icons' that 'yml' filetype should be ignored during "file" resolution which results in always relying on `vim.filetype.match()` output (which is essentially what is needed here). See [help entry](https://github.com/echasnovski/mini.nvim/blob/fb2cbeb4bff225e3c7907cfcd95ebf2a4492f8b6/doc/mini-icons.txt#L237) for more details.

Assuming the `vim.filetype.add({ pattern = { ['.*/roles/.*/tasks/.*%.ya?ml'] = 'yaml.ansible' } })` filetype detection set up, here is how it looks in latest 'mini.files':

![screenshot_2024-07-26_13:48:39](https://github.com/user-attachments/assets/1a6fee5c-bf9c-4a65-998f-705f3efef439)

Note that this requires that `MiniIcons.get('file', ...)` is called with a string that satisfies the pattern matching rules, which is usually a full path. So this might not work if 'mini.icons' user (usually plugin) calls it only with basename.

------

This question/issue was the main driving force for making those (overall positive) changes happen, so thanks for creating it!

**Reply by mstinsky** [reply]
@echasnovski great to hear, that is awesome, thank you very much for your work!

Sadly I somehow can't get it to work :/
I tested the following in my lazyvim config:
```
local ext_skip = { yml = true, yaml = true }

return {
  "echasnovski/mini.icons",
  opts = {
    filetype = {
      ['yaml.ansible'] = { glyph = '󱂚', hl = 'MiniIconsBlack' },
    },
    use_file_extension = function(ext, _) return not ext_skip[ext:lower()] end,
  }
}
```
It is my understanding that nothing more should be required.

My plugin version of mini.icons points to commit `842e51b` which includes the required changes if I am not mistaken.

Is there anything obvious to you what I might be doing wrong?

**Reply by echasnovski** [reply]
> Sadly I somehow can't get it to work :/

How do you test that? If through external plugin, make sure that it has support for 'mini.icons' and uses full paths in `MiniIcons.get()`. This is the case in latest 'mini.files'.

The best check is to test `MiniIcons.get()` directly:
- `:=MiniIcons.get('file', '/home/user/roles/a/tasks/hello.yml')` should print Ansible icon data.
- `:=MiniIcons.get('file', 'hello.yml') should print yaml icon data.

**Reply by mstinsky** [reply]
> The best check is to test `MiniIcons.get()` directly:
> 
> * `:=MiniIcons.get('file', '/home/user/roles/a/tasks/hello.yml')` should print Ansible icon data.
> * `:=MiniIcons.get('file', 'hello.yml') should print yaml icon data.

That helped, so it does work when calling `MiniIcons.get` with a test path. But in neo-tree.nvim where I try to get this to work it seems to depend on where I open neovim from.

Project root:
![Screenshot_20240727_160653](https://github.com/user-attachments/assets/130f0fbc-2d5d-485f-a44f-ff288ad4056c)

From withing the tasks folder of the project:
![Screenshot_20240727_160732](https://github.com/user-attachments/assets/8a500f5a-d5af-4f77-8ff9-b446044338d4)


So it seems neo-tree.nvim is calling mini.icons with different path? Not sure how all the components are interacting with each other.
Ill need to dig around a bit to see if I can get it to work somehow.

**Reply by echasnovski** [reply]
> So it seems neo-tree.nvim is calling mini.icons with different path? Not sure how all the components are interacting with each other.
> Ill need to dig around a bit to see if I can get it to work somehow.

'nvim-neo-tree/neo-tree.nvim' doe not yet have direct support for 'mini.icons'. See nvim-neo-tree/neo-tree.nvim#1510 and nvim-neo-tree/neo-tree.nvim#1527. So what you currently see is the effect of mocked 'nvim-web-devicons' module via [`MiniIcons.mock_nvim_web_devicons()`](https://github.com/echasnovski/mini.nvim/blob/e24ec1fcf885990fec1a993fe8304a033fd1806d/doc/mini-icons.txt#L440).

What I'd assume to happen is indeed that in both cases `MiniIcons.get('file', 'deploy.yml')` is called and in both cases is treated as relative path to the root. Hence the first case does not match (as its full paths is assumed to be '~/ws/marc/ansible/control/deploy.yml') and the second case it does match.

All in all, for take proper advantage of this feature, `get()` *must* be called with full path. Like it is done in 'mini.files' :wink:

**Comment by echasnovski** [comment]
@mstinsky, there was a slight change in how `use_file_extension` operates. It is now explicitly called once, even if extension is compound or file has dots. Plus extension is now converted to lowercase (as this is what is needed in 99% of cases while the second argument with the file name is supplied as is).

The approach with `ext_skip` and `return not ext_skip[ext:lower()]` will still work in most cases when there is only single dot in a basename (like 'roles/a/tasks/hello.yml'. However, it fails on rare cases like 'roles/a/tasks/hello.extradot.yml'.

The [suggested way](https://github.com/echasnovski/mini.nvim/blob/59ecf78491f29a9ec31f4124f3da77091f37e4d9/doc/mini-icons.txt#L265) now is to perform explicit string manipulation before comparing. So in your case it would be something like `use_file_extension = function(ext) return ext:sub(-3) ~= 'yml' and ext:sub(-4) ~= 'yaml' end`.
